connmgr/connmanager.go

```diff
@@ -0,0 +1,321 @@
+	// requests. Defaults to 1s.
+	RetryDuration time.Duration
+
+	// OnConnection is a callback that is fired when a new connection is
+	// established.
+	OnConnection OnConnectionFunc
+
+	// OnDisconnection is a callback that is fired when a connection is
+	// disconnected.
+	OnDisconnection OnDisconnectionFunc
+
+	// GetNewAddress is a way to get an address to make a network connection
+	// to.  If nil, no new connections will be made automatically.
+	GetNewAddress AddressFunc
+
+	// Dial connects to the address on the named network. It cannot be nil.
+	Dial DialFunc
+}
+
+// handleConnected is used to queue a successful connection.
+type handleConnected struct {
+	c    *ConnReq
+	conn net.Conn
+}
+
+// handleDisconnected is used to remove a connection.
+type handleDisconnected struct {
+	id    uint64
+	retry bool
+}
+
+// handleFailed is used to remove a pending connection.
+type handleFailed struct {
+	c   *ConnReq
+	err error
+}
+
+// ConnManager provides a manager to handle network connections.
+type ConnManager struct {
+	cfg            Config
+	wg             sync.WaitGroup
+	conns          map[uint64]*ConnReq
+	connReqCount   uint64
+	failedAttempts uint64
+	requests       chan interface{}
+	stop           int32
+	quit           chan struct{}
+}
+
+// handleFailedConn handles a connection failed due to a disconnect or any
+// other failure. If permanent, it retries the connection after the configured
+// retry duration. Otherwise, if required, it makes a new connection request.
+// After maxFailedConnectionAttempts new connections will be retried after the
+// configured retry duration.
+func (cm *ConnManager) handleFailedConn(c *ConnReq, retry bool) {
+	if atomic.LoadInt32(&cm.stop) != 0 {
+		return
+	}
+	if retry && c.Permanent {
+		c.retryCount++
+		d := time.Duration(c.retryCount) * cm.cfg.RetryDuration
+		if d > maxRetryDuration {
+			d = maxRetryDuration
+		}
+		log.Debugf("Retrying connection to %v in: %v", c.Addr, d)
+		time.AfterFunc(d, func() {
+			cm.Connect(c)
+		})
+	} else if cm.cfg.GetNewAddress != nil {
+		cm.failedAttempts++
+		if cm.failedAttempts > maxFailedAttempts {
+			log.Debugf("Max failed connection attempts reached: [%d] "+
+				"-- retrying connection in: %v", maxFailedAttempts,
+				cm.cfg.RetryDuration)
+			time.AfterFunc(cm.cfg.RetryDuration, func() {
+				cm.Connect(&ConnReq{Addr: cm.cfg.GetNewAddress()})
+			})
+		} else {
+			go cm.Connect(&ConnReq{Addr: cm.cfg.GetNewAddress()})
+		}
+	}
+}
+
+// connHandler handles all connection related requests.  It must be run as a
+// goroutine.
+//
+// The connection handler makes sure that we maintain a pool of active outbound
+// connections so that we remain connected to the network.  Connection requests
+// are processed and mapped by their assigned ids.
+func (cm *ConnManager) connHandler() {
+out:
+	for {
+		select {
+		case req := <-cm.requests:
+			switch msg := req.(type) {
+
+			case handleConnected:
+				connReq := msg.c
+				connReq.stateMtx.Lock()
```

Comment:
Since changing the state requires the mutex and it's duplicated in several places, I would suggest an `UpdateState` function on the `connReq` type like so:

```Go
// UpdateState changes the connection state to the provided state.
func (c *ConnReq) UpdateState(state ConnState) {
	c.stateMtx.Lock()
	c.state = state
	c.stateMtx.Unlock()
}
```

Then, this would just be:

```Go
connReq.UpdateState(ConnEstablished)
```

The same thing goes for the other places the connection request state is updated.